base/parallel/macros.jl

```diff
@@ -98,42 +98,160 @@ end
     each = div(N,np)
     extras = rem(N,np)
     nchunks = each > 0 ? np : extras
-    chunks = Array{UnitRange{Int}}(nchunks)
+    chunks = Dict{Int, UnitRange{Int}}()
     lo = 1
     for i in 1:nchunks
         hi = lo + each - 1
         if extras > 0
             hi += 1
             extras -= 1
         end
-        chunks[i] = lo:hi
+        chunks[wlist[i]] = lo:hi
         lo = hi+1
     end
     return chunks
 end
 
-function preduce(reducer, f, R)
-    N = length(R)
-    chunks = splitrange(N, nworkers())
-    all_w = workers()[1:length(chunks)]
+"""
+    ParallelAccumulator(initial)
+
+Constructs a reducing accumulator designed to be used in conjunction with `@parallel`
+for-loops. Takes a single argument specifying an initial value.
+
+The body of the `@parallel` for-loop can refer to multiple `ParallelAccumulator`s.
+The 0-arg indexation syntax, `[]` is used to fetch from and assign to an accumulator in the loop body.
+For example `acc[] = acc[] + i` will fetch the current, locally accumulated value in `acc`, add
+the value of `i` and set the new accumulated value back in `acc`.
+
+See [`@parallel`](@ref) for more details.
+
+ParallelAccumulators can also be used independent of `@parallel` loops.
+
+```julia
+acc = ParallelAccumulator(1)
+@sync for p in workers()
+    @spawnat p begin
+        for i in 1:10
+            acc[] += 1       # Local accumulation on each worker
+        end
+        push!(acc)           # Explicit push of local accumulation to driver node (typically node 1)
+    end
+end
+reduce(+, acc)
+```
+
+Usage of ParallelAccumulators independent of a `@parallel` construct must observe the following:
+- All remote tasks must be completed before calling `reduce` to retrieve the accumulated value. In the
+  example above, this is achieved by [`@sync`](@ref).
+- `push!(acc)` must be explictly called once on each worker. This pushes the locally accumulated value
+  to the node driving the computation.
+
+Note that the optional `initial` value is used on all workers. For example, if the reducing function is `+`,
+`ParallelAccumulator(25)` will add a total of `25*nworkers()` to the final result.
+"""
+type ParallelAccumulator{T}
+    initial::T
+    value::T
+    workers::Set{Int}       # Used on caller to detect arrival of all parts
+    chnl::RemoteChannel
+    hval::Int               # change hash value to ensure globals are serialized everytime
+
+    ParallelAccumulator(initial, chnl) = new(initial, initial, Set{Int}(), chnl, 0)
+end
+
+ParallelAccumulator{T}() where T  = ParallelAccumulator{T}(zero(T))
+ParallelAccumulator{T}(initial::T) = ParallelAccumulator{T}(initial, RemoteChannel(()->Channel{Tuple}(Inf)))
+
+const pacc_registry=Dict{RRID, Array{ParallelAccumulator}}()
+
+getindex(pacc::ParallelAccumulator) = pacc.value
+setindex!(pacc::ParallelAccumulator, v) = (pacc.value = v)
+
+hash(pacc::ParallelAccumulator, h::UInt) = hash(pacc.hval, hash(pacc.chnl, h))
+
+
+"""
+    push!(pacc::ParallelAccumulator)
```

Comment:
1-arg `push!` doesn't make sense from an API standpoint. This has more to do with the communication than a collection operation.